File: 34-find-first-and-last-position-of-element-in-sorted-array/find-first-and-last-position-of-element-in-sorted-array.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var searchRange = function(nums, target) {
    const lowerBound = lowerBoundBinarySearch(nums, target)
    const upperBound = upperBoundBinarySearch(nums, target)

    return [lowerBound, upperBound]
};

function lowerBoundBinarySearch(nums, target) {
    let left = 0
    let right = nums.length - 1

    while (left < right) {
        const mid = Math.floor((left + right) / 2)

        if (nums[mid] > target) {
            right = mid - 1
        } else if (nums[mid] < target) {
            left = mid + 1
        } else {
            right = mid
        }
    }

    return nums[left] === target ? left : -1
}

function upperBoundBinarySearch(nums, target) {
    let left = 0
    let right = nums.length - 1

    while (left < right) {
        const mid = Math.floor((left + right) / 2) + 1

        if (nums[mid] > target) {
            right = mid - 1
        } else if (nums[mid] < target) {
            left = mid + 1
        } else {
            left = mid
        }
    }

    return nums[right] === target ? right : -1
}

File: 34-find-first-and-last-position-of-element-in-sorted-array/find-first-and-last-position-of-element-in-sorted-array.py
class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        lower = self.lower(nums, target)
        higher = self.higher(nums, target)

        return [lower, higher]

    def lower(self, nums, target):
        l, r = 0, len(nums) - 1
        res = -1

        while l <= r:
            mid = (l + r) // 2
            if nums[mid] == target:
                res = mid
                r = mid - 1
            elif nums[mid] < target:
                l = mid + 1
            else:
                r = mid - 1
        return res

    def higher(self, nums, target):
        l, r = 0, len(nums) - 1
        res = -1

        while l <= r:
            mid = (l + r) // 2
            if nums[mid] == target:
                res = mid
                l = mid + 1
            elif nums[mid] < target:
                l = mid + 1
            else:
                r = mid - 1
        
        return res